src/QuikGraph/Algorithms/Observers/EdgePredecessorRecorderObserver.cs

```diff
@@ -188,7 +196,44 @@ private void OnEdgeFinished([NotNull] TEdge finishedEdge)
                     return;
             }
 
-            EndPathEdges.Add(finishedEdge);
+            _endPathEdges.Add(finishedEdge);
+        }
+
+        /// <summary>
+        /// Scope for disposing an <see cref="IObserver{TAlgorithm}"/> attachment without allocating closures or boxed
+        /// objects on the heap. You won't need to explicitly instantiate or use this <see langword="struct"/>, use the
+        /// enclosing observer class's <see cref="IObserver{TAlgorithm}.Attach"/> method in a <see langword="using"/>
+        /// statement instead.
+        /// </summary>
+        public struct AttachScope : IDisposable
```

Comment:
Making classes for these scenarios will certainly allow to keep those implementations details private which is better.